internal/controlplane/handlers_oauth.go

```diff
@@ -470,6 +473,47 @@ func (s *Server) VerifyProviderTokenFrom(ctx context.Context,
 	return &pb.VerifyProviderTokenFromResponse{Status: "OK"}, nil
 }
 
+// VerifyProviderCredential verifies the provider credential has been created for the matching enrollment nonce
+func (s *Server) VerifyProviderCredential(ctx context.Context,
+	in *pb.VerifyProviderCredentialRequest) (*pb.VerifyProviderCredentialResponse, error) {
+	entityCtx := engine.EntityFromContext(ctx)
+	projectID := entityCtx.Project.ID
+
+	enrollmentNonce := in.EnrollmentNonce
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Project = projectID
+
+	// check if a token has been created matching the nonce
+	_, err := s.store.GetAccessTokenByEnrollmentNonce(ctx, db.GetAccessTokenByEnrollmentNonceParams{
+		ProjectID:       projectID,
+		EnrollmentNonce: sql.NullString{Valid: true, String: enrollmentNonce},
+	})
+
+	if errors.Is(err, sql.ErrNoRows) {
+		// if there's no token, check for an installation ID
+		_, err = s.store.GetInstallationIDByEnrollmentNonce(ctx,
+			db.GetInstallationIDByEnrollmentNonceParams{
+				ProjectID: uuid.NullUUID{
+					Valid: true,
+					UUID:  projectID,
+				},
+				EnrollmentNonce: sql.NullString{Valid: true, String: enrollmentNonce},
+			},
+		)
+
+		if errors.Is(err, sql.ErrNoRows) {
+			return &pb.VerifyProviderCredentialResponse{Status: "KO"}, nil
+		} else if err != nil {
+			return nil, status.Errorf(codes.Internal, "error getting installation ID: %v", err)
+		}
+	} else if err != nil {
```

Comment:
nit: for readability, let's move error handling to the top. This function is quite indented that it gets hard to read.

Reply:
I moved the `if` blocks around, hopefully it's more readable now, but let me know if you see a better ordering of the `if` statements.

Reply:
I put in a suggestion to have two early-exit cases (found successful PAT and found successful install), and then have the default case to be return `KO`.